upgrade-agent/server.go

```diff
@@ -0,0 +1,140 @@
+
+	server.grpcServer = grpcServer
+	return server, nil
+}
+
+// Run starts the upgrade-agent server on the given port, using the provided protocol and socket address.
+func (s *Server) Run(protocol string, sockAddr string) error {
+	grpcServer := grpc.NewServer()
+
+	upgradeproto.RegisterUpdateServer(grpcServer, s)
+
+	cleanup := func() error {
+	err := os.RemoveAll(sockAddr)
+	if errors.Is(err, fs.ErrNotExist) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+		return nil
+	}
+	if err := cleanup(); err != nil {
+		return fmt.Errorf("failed to clean socket file: %s", err)
+	}
+
+	lis, err := net.Listen(protocol, sockAddr)
+	if err != nil {
+		return fmt.Errorf("failed to listen: %s", err)
+	}
+
+	s.log.Infof("Starting")
+	return grpcServer.Serve(lis)
+}
+
+// Stop stops the upgrade-agent server gracefully.
+func (s *Server) Stop() {
+	s.log.Infof("Stopping")
+
+	s.grpcServer.GracefulStop()
+
+	s.log.Infof("Stopped")
+}
+
+// ExecuteUpdate installs & verifies the provided kubeadm, then executes `kubeadm upgrade plan` & `kubeadm upgrade apply {wanted_Kubernetes_Version}` to upgrade to the specified version.
+func (s *Server) ExecuteUpdate(ctx context.Context, updateRequest *upgradeproto.ExecuteUpdateRequest) (*upgradeproto.ExecuteUpdateResponse, error) {
+	s.log.Infof("Upgrade to Kubernetes version started: %s", updateRequest.WantedKubernetesVersion)
+
+	installer := installer.NewOSInstaller()
+	err := installer.Install(ctx, versions.ComponentVersion{
+		URL:         updateRequest.KubeadmUrl,
+		Hash:        updateRequest.KubeadmHash,
+		InstallPath: constants.KubeadmPath,
+		Extract:     false,
+	})
+	if err != nil {
+		return &upgradeproto.ExecuteUpdateResponse{}, status.Errorf(codes.Internal, "unable to retrieve the specified kubeadm binary: %s", err)
+	}
+
+	err = verifyVersion(updateRequest.WantedKubernetesVersion)
+	if err != nil {
+		return &upgradeproto.ExecuteUpdateResponse{}, status.Errorf(codes.Internal, "unable to verify the Kubernetes version %s: %s", updateRequest.WantedKubernetesVersion, err)
+	}
+
+	upgradeCmd := exec.CommandContext(ctx, "kubeadm", "upgrade", "plan")
+	if err := upgradeCmd.Run(); err != nil {
+		return &upgradeproto.ExecuteUpdateResponse{}, status.Errorf(codes.Internal, "unable to execute kubeadm upgrade plan: %s", err)
+	}
+
+	applyCmd := exec.CommandContext(ctx, "kubeadm", "upgrade", "apply", updateRequest.WantedKubernetesVersion)
+	if err := applyCmd.Run(); err != nil {
+		return &upgradeproto.ExecuteUpdateResponse{}, status.Errorf(codes.Internal, "unable to execute kubeadm upgrade apply: %s", err)
+	}
+
+	s.log.Infof("Upgrade to Kubernetes version succeeded: %s", updateRequest.WantedKubernetesVersion)
+	return &upgradeproto.ExecuteUpdateResponse{}, nil
+}
+
+// verifyVersion verifies the provided Kubernetes version.
+func verifyVersion(version string) error {
+	if !versionRegexp.MatchString(version) {
+		return fmt.Errorf("invalid kubernetes version: %s", version)
+	}
+	return nil
+}
```

Comment:
I think the [semver package](https://pkg.go.dev/golang.org/x/mod/semver#IsValid) provides this functionality.